### ISSUE TITLE
VsTestPlatformToolInstaller from Artifacts feed fails with '401 Unauthorized'

### ISSUE THREAD

**Issue by NKnusperer** [Microsoft/azure-pipelines-tasks#9757]
## Environment
- Azure DevOps Server 2019 RTW (17.143.28621.4)
- Agent v2.147.1 inside Docker Container based on `microsoft/dotnet-framework:4.7.2-runtime-20190212-windowsservercore-ltsc2019`

## Issue Description

I'am trying to use the `VisualStudioTestPlatformInstaller@1` task with the `customFeed` set to a Artifacts feed configured as a Upstream source for nuget.org (same Project Collection):
```yaml
- task: VisualStudioTestPlatformInstaller@1
  displayName: "Visual Studio Test Platform Installer"
  inputs:
    packageFeedSelector: customFeed
    customFeed: "https://srv-ados/MyCompany/_packaging/product-artifacts/nuget/v3/index.json"
    versionSelector: specificVersion
    testPlatformVersion: 15.9.1
    password: $(System.AccessToken)
```

The task is failing because of `The remote server returned an error: (401) Unauthorized.`
Using a PAT with full rights instead of `$(System.AccessToken)` also has the same result.
### Task logs

[log_8_128.zip](https://github.com/Microsoft/azure-pipelines-tasks/files/2942495/log_8_128.zip)

### Error logs
```
Attempting to gather dependency information for package 'Microsoft.TestPlatform.15.9.1' with respect to project 'C:\BuildAgent\_work\_temp\VsTest', targeting 'Any,Version=v0.0'
Unable to load the service index for source https://srv-ados/MyCompany/_packaging/product-artifacts/nuget/v3/index.json.
  An error occurred while sending the request.
  The remote server returned an error: (401) Unauthorized.
  No credentials are available in the security package
##[error]Failed to download version 15.9.1 of the test platform with error Error: C:\BuildAgent\_work\_tasks\VisualStudioTestPlatformInstaller_2c65196a-54fd-4a02-9be8-d9d1837b7111\1.1.5\nuget.exe failed with return code: 1.
```

**Comment by ShreyasRmsft** [comment]
@NKnusperer lemme look into this.

**Comment by ShreyasRmsft** [comment]
![image](https://user-images.githubusercontent.com/25194518/54429794-84eebf80-4747-11e9-9a08-5eb360b7f3cc.png)

One thing to immediately check is if you have permission to the feed. One simple way to verify is to use the same PAT token and locally try to fetch the package from the private feed. Make sure the cached credentials don't get used and the provided PAT is used.

Can you tell me the result for this?

**Comment by ShreyasRmsft** [comment]
Also follow this guide to use PAT for authentication:

https://docs.microsoft.com/en-us/azure/devops/artifacts/nuget/nuget-exe?view=azure-devops#add-a-feed-to-nuget-2

I've pasted the relevant section here for you

![image](https://user-images.githubusercontent.com/25194518/54436372-65ac5e00-4758-11e9-9630-05c0b2d12875.png)

Lemme know if this works. PAT is supported with v2 feeds

**Comment by ShreyasRmsft** [comment]
Basically change your feed to https://srv-ados/MyCompany/_packaging/product-artifacts/nuget/v2

**Comment by NKnusperer** [comment]
Using `nuget.exe` command-line tool with a PAT on my local machine does not work either (sorry for German OS language):
```powershell
> .\nuget.exe sources Add -NonInteractive -Name product-artifacts -Source https://srv-ados/MyCompany/_packaging/product-artifacts/nuget/v3/index.json -Username vstestPlatformToolInstaller -Password {MY_PAT} -ConfigFile test.config -StorePasswordInClearText
> .\nuget.exe install Microsoft.TestPlatform -Version 15.9.1 -Source product-artifacts -OutputDirectory . -NoCache -DirectDownload -ConfigFile test.config -NonInteractive
Feeds used:
  https://srv-ados/MyCompany/_packaging/product-artifacts/nuget/v3/index.json



Attempting to gather dependency information for package 'Microsoft.TestPlatform.15.9.1' with respect to project '.', targeting 'Any,Version=v0.0'
MSBuild auto-detection: using msbuild version '15.9.21.664' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin'.
Using credentials from config. UserName: vstestPlatformToolInstaller
Unable to load the service index for source https://srv-ados/MyCompany/_packaging/product-artifacts/nuget/v3/index.json.
  Der Antwortstatuscode gibt keinen Erfolg an: 401 (Unauthorized).
```

`https://srv-ados/MyCompany/_packaging/product-artifacts/nuget/v2` as feed URL yields the same result.

Using the feed without explicit credentials as Windows Domain User Identity works.
Also the `NuGetCommand@2` task restores fine on the Build Agent using the `vstsFeed` input parameter.

**Comment by ShreyasRmsft** [comment]
@NKnusperer you are right it doesn't work even after changing to v2. I am investigating this and will get back to you with a fix.

**Comment by ShreyasRmsft** [comment]
@NKnusperer sorry for the wait.

Azure DevOps Server today tries to force the authentication type to be negotiate today even though you provide username and password (basic authentication)

In the Vstest tool installer we are creating a temporary config file and feeding in the pat token but this is not getting used because the server first tries to use negotiate auth and fails then and there, never falling back to use basic auth.

There were ongoing issues in the nuget repo where finally in nuget.exe version 5.0 they added an additional parameter -ValidAuthenticationTypes (value set to basic) which will configure nuget to reject negotiate and use basic auth.

I will work on incorporating this fix into the latest version of vstestplatform tool installer. 

But to unblock yourself i would suggest that you use the network share option to unblock yourself for now.

And thank you for bringing this to our notice 👍

**Comment by ShreyasRmsft** [comment]
@NKnusperer were you able to unblock youself?

**Comment by ShreyasRmsft** [comment]
@NKnusperer i made the fix in https://github.com/Microsoft/azure-pipelines-tasks/pull/9917 
The fix will be available in the coming onprem releases.

**Comment by aakobordenau** [comment]
Hi, this is @NKnusperer :
@ShreyasRmsft Still does not work with on-premises Azure DevOps 2019.0.1 (17.143.28912.1) and Agent v2.153.2.
Task "Visual Studio Test Platform Installer" is reported as version 1.1.5.

**Comment by ShreyasRmsft** [comment]
@aakobordenau here's the PR where i made the fix: #9917

The task version for the task where this is fixed is 1.150.0 (we changed the versioning scheme to match internal sprint number representation, hence the big jump in numbers if you were wondering).

I think this is getting shipped in Azure DevOps 2019.1RC1 for OnPrem.

**Comment by v-milindm** [comment]
Hello, sorry for bumping up on this old thread but I am facing this exact same issue if anyone can help me here. Let me know if I should create a new issue for this.

@ShreyasRmsft , is this issue fixed?

@NKnusperer / @aa-ko, were you able to fix this issue on your end?

I am having the exact same issue with my build pipeline. I tried couple of things from the above suggestions mentioned in the comments but no luck.

Here's my setup:
steps:
- task: VisualStudioTestPlatformInstaller@1
  displayName: 'Visual Studio Test Platform Installer'
  inputs:
    packageFeedSelector: customFeed
    versionSelector: latestStable
    customFeed: 'https://{base_url}/{organization_name}/{project_name}/_packaging/{package_name}/nuget/v2'
    password: '$(System.AccessToken)'

This is the error from the log:
The V2 feed at 'https://{base_url}/{organization}/{project_name}/_packaging/{package_name}/nuget/v2/Search()?$filter=IsLatestVersion&$orderby=Id&searchTerm='Microsoft.TestPlatform'&targetFramework=''&includePrerelease=false&$skip=0&$top=30&semVerLevel=2.0.0' returned an unexpected status code '401 Unauthorized'.
MSBuild auto-detection: using msbuild version '17.2.1.25201' from 'C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\bin'.
Using credentials from config. UserName: vstestPlatformToolInstaller
##[error]Nuget.exe returned error code: 1
##[error]Failed to fetch list of available packages from nuget, looking for latest stable version cached.
Error: Listing packages failed. Nuget.exe returned 1.
Standard Error:
The V2 feed at 'https://{base_url}/{organization}/{project_name}/_packaging/{package_name}/nuget/v2/Search()?$filter=IsLatestVersion&$orderby=Id&searchTerm='Microsoft.TestPlatform'&targetFramework=''&includePrerelease=false&$skip=0&$top=30&semVerLevel=2.0.0' returned an unexpected status code '401 Unauthorized'.
Standard Out:
MSBuild auto-detection: using msbuild version '17.2.1.25201' from 'C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\bin'.
Using credentials from config. UserName: vstestPlatformToolInstaller

##[error]Failed to find any stable version of the test platform package in the cache.
##[error]Error: Failed to acquire the test platform. Please set system.debug to true and refer to the debug logs for additional information.
Finishing: Visual Studio Test Platform Installer

Not sure, if I am missing something here but any help would be greatly appreciated.
Thank you!
